### PR TITLE
[processing] Port functionality from history dialog to history provider framework

### PR DESCRIPTION
**Pull request by nyalldawson**
This is a copy port for now, there's no actual changes to the existing code.

### CODE REVIEW

File: src/gui/processing/qgsprocessinghistoryprovider.h [hunk 1]
```diff
@@ -42,11 +44,32 @@ class GUI_EXPORT QgsProcessingHistoryProvider : public QgsAbstractHistoryProvide
      */
     void portOldLog();
 
+    QgsHistoryEntryNode *createNodeForEntry( const QgsHistoryEntry &entry, const QgsHistoryWidgetContext &context ) override SIP_FACTORY;
+
+  signals:
+
+    /**
+     * Emitted when the provider needs to execute python \a commands in the Processing context.
```
Comment (nirvn) [inline src/gui/processing/qgsprocessinghistoryprovider.h:52]: \since ?

File: src/gui/processing/qgsprocessinghistoryprovider.h [hunk 2]
```diff
@@ -42,11 +44,32 @@ class GUI_EXPORT QgsProcessingHistoryProvider : public QgsAbstractHistoryProvide
...
+     */
+    void executePython( const QString &commands );
+
+    /**
+     * Emitted when the provider needs to create a Processing test with the given python \a command.
```
Comment (nirvn) [inline src/gui/processing/qgsprocessinghistoryprovider.h:57]: \since ?

File: src/gui/processing/qgsprocessinghistoryprovider.h [hunk 3]
```diff
@@ -42,11 +44,32 @@ class GUI_EXPORT QgsProcessingHistoryProvider : public QgsAbstractHistoryProvide
...
+     */
+    void createTest( const QString &command );
+
   private:
 
+    //! Executes some python commands
+    void execute( const QString &commands );
+
+    void emitCreateTest( const QString &command );
```
Comment (nirvn) [inline src/gui/processing/qgsprocessinghistoryprovider.h:66]:
```suggestion
    void createTest( const QString &command );
```

?

File: src/gui/processing/qgsprocessinghistoryprovider.cpp [hunk 4]
```diff
@@ -72,8 +81,195 @@ void QgsProcessingHistoryProvider::portOldLog()
...
   const QString userDir = QgsApplication::qgisSettingsDirPath() + QStringLiteral( "/processing" );
   return userDir + QStringLiteral( "/processing.log" );
 }
+
+void QgsProcessingHistoryProvider::execute( const QString &commands )
+{
+  emit executePython( commands );
+}
+
+void QgsProcessingHistoryProvider::emitCreateTest( const QString &command )
```
Comment (nirvn) [inline src/gui/processing/qgsprocessinghistoryprovider.cpp:272]: I've raised it in the header file already but to be consistent here, we should either call this createTest, or rename execute to emitExecute.

I'd favor renaming this function to createTest.

**Review by nirvn — APPROVED** [review]

**Review by alexbruy — APPROVED** [review]

**Pull request merged**